### PR TITLE
libiomp 20150401

### PR DESCRIPTION
**Pull request by tschoonj**
version bump
header installation path changed to prefix/include

see also discussion in #38707

### CODE REVIEW

**Comment by xu-cheng** [comment]
I think you need to update the test case in `clang-omp` as well. Also, from the backward compatibility aspect, maybe create a symlink to the original place? Will let the other maintainers decide it.

**Comment by tschoonj** [comment]
Ok, done. Will create a pull-request for clang-omp as soon as this is accepted

**Comment by tschoonj** [comment]
@xu-cheng ...and the test failed because the clang-omp test doesn't work anymore.

Any suggestions? the pull request for clang-omp will also fail because it will still use the old libiomp...

**Comment by xu-cheng** [comment]
> Ok, done. Will create a pull-request for clang-omp as soon as this is accepted

You can update `clang-omp` in this PR.

**Comment by tschoonj** [comment]
Nice, it worked!

**Comment by tdsmith** [comment]
If I want to build something against OpenMPI with gcc while I have clang-omp installed, does having this header directly in HOMEBREW_PREFIX/include cause trouble for my gcc build? Ping @metacollin.

**Comment by iml** [comment]
@tdsmith I haven't tested it out yet but it looks like only bin is linked, the headers are in libexec.

**Comment by tschoonj** [comment]
@xu-cheng How about pulling this one in? Thanks!

**Comment by xu-cheng** [comment]
@tschoonj 

* I would like to create symlink in old directory to keep backward compatibility. 
* I'm not quite familiar to OpenMP and its relationship with different compilers. So will let @tdsmith decide whether we should merge this.

**Comment by tdsmith** [comment]
Oh, I see! This looks okay; I didn't notice the `-I#{include}` in the test. (Sorry, I've been afk for the last week.)

**Comment by tdsmith** [comment]
@BrewTestBot test this please

**Comment by xu-cheng** [comment]
Ping @tdsmith

**Comment by tdsmith** [comment]
Oh, I let the damn build expire again, shame on me.

@BrewTestBot test this please

**Comment by xu-cheng** [comment]
@BrewTestBot test this please

**Comment by xu-cheng** [comment]
Ping @tdsmith in case you forget again. :smiley:

**Comment by tdsmith** [comment]
Thanks :sweat_smile: 

This _does_ install a `HOMEBREW_PREFIX/include/omp.h`. It does not cause problems with gcc's omp implementation in the default prefix, because:

```
#include <...> search starts here:
 /usr/local/Cellar/gcc/4.9.2_1/lib/gcc/4.9/gcc/x86_64-apple-darwin14.0.0/4.9.2/include <-- omp.h is here
 /usr/local/include
 ...
```

and `-I/usr/local/include` is always ignored. But if I add e.g. `-I/Users/tim/.brew/include`, it will not be ignored, and the libiomp omp.h will be found instead of gcc's omp.h.

I would like to not put this omp.h in the default search path. Can we teach clang-omp an additional default header search path?

**Comment by xu-cheng** [comment]
I think there is a talking that we should vendor libiomp inside clang-omp and boneyard this formula. It seems libiomp is only used by clang-omp. See #38707.

**Comment by xu-cheng** [comment]
Also FWIW, I want to have clang-omp supported in homebrew compiler list. So if it a formula`need :openmp`, we will choose clang-omp instead of gcc to avoid libc++ and libstdc++ incompatible problem.

**Comment by metacollin** [comment]
Sorry, been busy :(.  

Custom include paths must be added by setting the precompiler define C_INCLUDE_DIRS to the paths.  C_INCLUDE_DIRS is searched after the default search paths, so we can simply add whereever libiomp's include directory lives.  One can find the actual #define near the top of llvm/tools/clang/include/clang/Config/config.h (It's the second or third define I believe). 

Also, according to http://clang.llvm.org/get_started.html , cmake is the standard and preferred build system (I am not sure if this changed recently or its said that for a while and I never noticed heh).  The cmakelists for clang will safely insert whatever we want into the config.h file, we simply pass `-DC_INCLUDE_DIRS="#{homebrew_path_to_extra_include_dirs}"` when calling cmake, and we're good to go.  I've tested this and it works correctly.  I am actually not too familiar with autotools, so if anyone knows a safe way to set that define without switching to cmake, the define we want is C_INCLUDE_DIRS.  

Otherwise, we might consider switching the build system to cmake, which shouldn't be that big a problem.

Also, my personal opinion is that libiomp should remain a separate formula.  It's a runtime library, one does not actually need it to build clang-omp, or even compile OpenMP aware programs, it is only needed to execute them.  Intel releases new versions of the runtime fairly often, often with bug fixes and performance improvements, and it's an open OpenMP API runtime project, and can be used by any compiler, and should be a separate formula if only to make it available to developers using homebrew that are interested in adding OpenMP support to a compiler.  And there is potential for other things to use the runtime, even if the only forumla currently using it is clang-omp.  

But that's just my opinion, either way it's not going to be terrible if its separate, or terrible if its rolled in as a resource block in the clang-omp formula either.  Just my two satoshis on the subject.

**Comment by metacollin** [comment]
D'oh, its done with `--with-c-include-dirs=`.  :neutral_face: :palm_tree:  (face palm).  

Forget about the cmake stuff heh.  For being so easy, it sure isn't documented anywhere in the clang online docs.  I discovered it looking through the configure script.  

I am going to do a pull request momentarily, I'm waiting for clang to rebuild to double check that this actually does what it's supposed to.  I'm on a 2008 Mac Pro, so clang takes a while to build =/.

**Comment by tdsmith** [comment]
Neat; `Formula["libiomp"].opt_include/"libiomp"` is probably the path we want to add.

Is there a corresponding option for linking libiomp5.dylib?

**Comment by metacollin** [comment]
tdsmith, yep, that's what I am testing as we speak! 

As for linking libiomp5, the name is unique, nothing will link to it unless it means to.  If it were called libomp, it might be a problem, but there is one and only one library with the 'i' in front of the omp, and that's intel's runtime, with i signifying intel.  Regardless, I am looking for info on how to add a custom lib search path to clang, but haven't found anything yet.

**Comment by tdsmith** [comment]
That makes sense; let's not worry about hiding the library. I think it could be helpful if Homebrew is installed to a prefix other than /usr/local but people who do that should know how to use LDFLAGS.

**Comment by xu-cheng** [comment]
`clang-omp` has been updated. @tschoonj Please update this accordingly. Thanks.

**Comment by tschoonj** [comment]
Nothing needs an update I think. I just tested it on my machine with the new clang-omp. Perhaps trigger a rebuild and see what happens?

**Comment by xu-cheng** [comment]
You need to rebase and remove the head file installation location change

**Comment by tschoonj** [comment]
this ok?

File: Library/Formula/libiomp.rb [hunk 1]
```diff
@@ -35,7 +35,7 @@ def install
     system "cmake", ".", *args
     system "make", "all", "common"
 
-    (include/"libiomp").install Dir["exports/common/include/*"]
```
Comment (xu-cheng) [inline Library/Formula/libiomp.rb]: Don't change this. It should be installed in `include/libiomp` now.

File: Library/Formula/libiomp.rb [hunk 2]
```diff
@@ -51,7 +51,7 @@ def install
       }
     EOS
     (testpath/"test.c").write(testfile)
-    system ENV.cc, "-L#{lib}", "-liomp5", "-I#{include}/libiomp", "test.c", "-o", "test"
```
Comment (xu-cheng) [inline Library/Formula/libiomp.rb]: Don't change this line, either.

**Comment by metacollin** [comment]
FWIW, clang-omp stopped building for me too before this commit, but there had been no change to the formula.  It looked like ranlib was no longer getting called correctly by the Makefile.  @xu-cheng mentioned the llvm formula had to move to cmake recently too, did autotools get updated recently that broke things? It broke for the test-bot too I believe.

**Comment by tschoonj** [comment]
Ok, done

**Comment by metacollin** [comment]
Test-bot should tell me if I'm crazy or not at least :)

**Comment by xu-cheng** [comment]
Thanks for the pull request! 🎉 Homebrew depends on contributions from community members like you and we're grateful for your support.

**Pull request closed without merging**